File: app/src/main/java/com/sdin/tourstamprally/ui/fragment/VisitHistoryFragment.java
package com.sdin.tourstamprally.ui.fragment;

import android.annotation.SuppressLint;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.databinding.DataBindingUtil;
import androidx.navigation.Navigation;
import androidx.recyclerview.widget.RecyclerView;

import com.bumptech.glide.Glide;
import com.sdin.tourstamprally.R;
import com.sdin.tourstamprally.Utils;
import com.sdin.tourstamprally.adapter.ViSitReAdapter;
import com.sdin.tourstamprally.databinding.DirectionGuidTagItemBinding;
import com.sdin.tourstamprally.databinding.FragmentVisithistoryBinding;
import com.sdin.tourstamprally.model.CouponModel;
import com.sdin.tourstamprally.model.ReviewWriter;
import com.sdin.tourstamprally.model.history_spotModel2;
import com.sdin.tourstamprally.ui.dialog.PopUp_Image;
import com.sdin.tourstamprally.utill.listener.ItemCliclListener;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers;
import io.reactivex.rxjava3.observers.DisposableSingleObserver;
import io.reactivex.rxjava3.schedulers.Schedulers;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class VisitHistoryFragment extends BaseFragment {

    private FragmentVisithistoryBinding binding;
    private List<history_spotModel2> history_spotList;
    private ViSitReAdapter adapter;


    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_visithistory, container, false);
        Log.wtf("VisitHistoryFragment", "onCreateView");
        return binding.getRoot();
    }


    @SuppressLint("ClickableViewAccessibility")
    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        binding.progressBar.setVisibility(View.VISIBLE);
        setVisitInit();
        setCatecory();

        binding.searchEdt.addTextChangedListener(new TextWatcher() {

            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                search(binding.searchEdt.getText().toString());
            }

        });

        binding.searchBtn.setOnClickListener(v -> search(binding.searchEdt.getText().toString()));


    }

    private void insert_intest(int touristspot_idx) {
        apiService.insert_intest(Utils.User_Idx, touristspot_idx).enqueue(new Callback<Integer>() {
            @Override
            public void onResponse(@NotNull Call<Integer> call, @NotNull Response<Integer> response) {
                if (response.isSuccessful()) {
                    ShowToast("찜하기에 성공 하셨습니다.", requireContext());
                } else {
                    //Log.wtf("찜하기 실패", response.toString());
                    ShowToast("찜하기에 실패 하셨습니다.", requireContext());
                }
            }

            @Override
            public void onFailure(@NotNull Call<Integer> call, @NotNull Throwable t) {
                t.printStackTrace();
            }
        });
    }

    private void setVisitInit() {
        /*apiService.getHistorySpot(Utils.User_Idx).enqueue(new Callback<List<history_spotModel2>>() {
            @SuppressLint("ClickableViewAccessibility")
            @Override
            public void onResponse(@NotNull Call<List<history_spotModel2>> call, @NotNull Response<List<history_spotModel2>> response) {
                binding.progressBar.setVisibility(View.GONE);
                try {

                    if (response.isSuccessful()) {

                        if (response.body() != null) {

                            history_spotList = new ArrayList<>(response.body());

                            SwipeHelperCallback callback = new SwipeHelperCallback();
                            callback.setClamp(-200f);
                            ItemTouchHelper itemTouchHelper = new ItemTouchHelper(callback);
                            itemTouchHelper.attachToRecyclerView(binding.visitRecyclerView);
                            ArrayList<history_spotModel2> arrayList = new ArrayList<>(history_spotList);
                            adapter = new ViSitReAdapterV2(arrayList, requireContext(), model -> {
                                onWriteReviewClick(model);
                            });

                            binding.visitRecyclerView.setLayoutManager(new LinearLayoutManager(requireContext()));
                            binding.visitRecyclerView.setAdapter(adapter);
                            binding.visitRecyclerView.addItemDecoration(new ItemDecoration());
                            binding.visitRecyclerView.setOnTouchListener((v, event) -> {
                                //  swipeHelperCallback2.removePreviousClamp(binding.visitRecyclerView);
                                callback.removePreviousClamp(binding.visitRecyclerView);
                                return false;
                            });
                            setRecyclerView();

                        }


                    }
                } catch (Exception e) {
                    e.printStackTrace();
                }
            }

            @Override
            public void onFailure(@NotNull Call<List<history_spotModel2>> call, @NotNull Throwable t) {
                t.printStackTrace();
            }
        });*/
    }

    private void onWriteReviewClick(@NonNull ReviewWriter model) {
        //setFragment("리뷰작성", WriteReviewFragment.newInstance(reviewWriter));
        //Navigation.findNavController().navigate(R.id.);
        Bundle bundle = new Bundle();
        bundle.putSerializable("model", model);
        Navigation.findNavController(requireActivity(), R.id.nav_host)
                .navigate(R.id.action_fragment_visithistory_to_fragment_write_review, bundle);
    }

    private void setRecyclerView() {
        adapter.itemClickListener(new ItemCliclListener() {
            @Override
            public void deapsClick(int position, history_spotModel2 model) {
                apiService.select_interest_status(Utils.User_Idx, model.getTouristspot_idx()).enqueue(new Callback<Integer>() {
                    @Override
                    public void onResponse(@NotNull Call<Integer> call, @NotNull Response<Integer> response) {
                        if (response.isSuccessful()) {
                            //ShowToast("찜하기에 성공 하셨습니다.", requireContext());
                            //Log.wtf("response.body()", String.valueOf(response.body()));

                            if (response.body() != null && response.body() == 0) {
                                //  Log.wtf("onResponse", "찜하기 000000");
                                insert_intest(model.getTouristspot_idx());
                            }

                        } else {
                            //Log.wtf("찜하기 실패", response.toString());
                            ShowToast("찜하기에 실패 하셨습니다.", requireContext());
                        }
                    }

                    @Override
                    public void onFailure(@NotNull Call<Integer> call, @NotNull Throwable t) {
                        ShowToast("찜하기에 실패 하셨습니다.", requireContext());
                        t.printStackTrace();
                    }

                });
            }

            @Override
            public void delReview(int review_idx, int position) {
                //리뷰 삭제
                apiService.deleteReview(review_idx).subscribeOn(Schedulers.newThread())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(new DisposableSingleObserver<Integer>() {
                            @Override
                            public void onSuccess(@io.reactivex.rxjava3.annotations.NonNull Integer integer) {
                                if (integer > 0) {
                                    adapter.setChange(position);
                                    Toast.makeText(requireContext(), "리뷰 삭제에 성공하였습니다.", Toast.LENGTH_SHORT).show();
                                }

                            }

                            @Override
                            public void onError(@io.reactivex.rxjava3.annotations.NonNull Throwable e) {
                                e.printStackTrace();
                            }
                        });
            }

            @Override
            public void clearClick(int touristspot_idx) {
                apiService.selectCoupon(touristspot_idx, Utils.User_Idx).subscribeOn(Schedulers.newThread())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(new DisposableSingleObserver<CouponModel>() {
                            @Override
                            public void onSuccess(@io.reactivex.rxjava3.annotations.NonNull CouponModel couponModel) {

                                try {
                                    if (couponModel != null) {
                                        new PopUp_Image(requireContext(), couponModel).show();
                                    } else {
                                        Toast.makeText(requireContext(), "해당 관광지의 쿠폰이 발급되지 않았습니다.", Toast.LENGTH_SHORT).show();
                                    }
                                } catch (Exception e) {
                                    e.printStackTrace();
                                    Toast.makeText(requireContext(), "해당 관광지의 쿠폰이 발급되지 않았습니다.", Toast.LENGTH_SHORT).show();
                                }


                            }

                            @Override
                            public void onError(@io.reactivex.rxjava3.annotations.NonNull Throwable e) {
                                e.printStackTrace();
                                Toast.makeText(requireContext(), "해당 관광지의 쿠폰이 발급되지 않았습니다.", Toast.LENGTH_SHORT).show();
                            }
                        });
            }

        });
    }


    private void setCatecory() {

        CaterotyAdaper caterotyAdaper = new CaterotyAdaper(new ArrayList<>(Arrays.asList(requireContext().getResources().getStringArray(R.array.area_direction))));
        binding.categoryRecyclerView.setAdapter(caterotyAdaper);
        caterotyAdaper.itemOnClick = location -> {
            if (location.equals("전체")) {
                location = "";
            }
            search(location);
        };
    }


    @SuppressLint("NotifyDataSetChanged")
    private void search(String searchData) {

        ArrayList<history_spotModel2> arrayList = new ArrayList<>();
        if (searchData.length() == 0) {

            arrayList.addAll(history_spotList);

        } else {

            for (int i = 0; i < history_spotList.size(); i++) {
                if (history_spotList.get(i).getLocation_name().toLowerCase().contains(searchData)) {
                    arrayList.add(history_spotList.get(i));
                }

                if (history_spotList.get(i).getTouristspot_name().toLowerCase().contains(searchData)) {
                    arrayList.add(history_spotList.get(i));
                }

            }
        }

        adapter.setList(arrayList);
        adapter.notifyDataSetChanged();
    }


    interface ItemOnClick {
        void onCategoryClick(String location);
    }

    private class CaterotyAdaper extends RecyclerView.Adapter<CaterotyAdaper.ViewHolder> {

        private final ArrayList<String> arrayList;

        private ItemOnClick itemOnClick;

        private int selectedItem = 0;
        private int prevSelected = -1;

        public CaterotyAdaper(ArrayList<String> arrayList) {
            this.arrayList = arrayList;
        }

        @NonNull
        @Override
        public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            DirectionGuidTagItemBinding binding = DirectionGuidTagItemBinding.inflate(LayoutInflater.from(parent.getContext()), parent, false);
            return new ViewHolder(binding);
        }

        @Override
        public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
            holder.binding.tagItemTxv.setText(arrayList.get(position));
            setBackground(position, holder);
        }


        private void setBackground(int position, CaterotyAdaper.ViewHolder holder) {
            int icon;
            if (selectedItem == position) {
                holder.binding.tagItemContainer.setBackground(ContextCompat.getDrawable(holder.binding.tagItemContainer.getContext(), R.drawable.bg_rounded_category_selected));
                holder.binding.tagItemTxv.setTextColor(ContextCompat.getColor(holder.binding.tagItemTxv.getContext(), R.color.white));
                icon = setIcon(true, position);
                Log.wtf("onBindViewHolder", "setBackground if ");
                /*if(position == 0){
                    holder.binding.tagItemImg.setImageResource(R.drawable.ic_roadtour_on);
                }*/


            } else {
                holder.binding.tagItemContainer.setBackground(ContextCompat.getDrawable(holder.binding.tagItemContainer.getContext(), R.drawable.bg_rounded_category));
                holder.binding.tagItemTxv.setTextColor(ContextCompat.getColor(holder.binding.tagItemTxv.getContext(), R.color.mainColor));
                icon = setIcon(false, position);
                Log.wtf("onBindViewHolder", "setBackground elseA ");
                /*if(position == 0){
                    holder.binding.tagItemImg.setImageResource(R.drawable.ic_roadtour_off);
                }*/

            }

            Glide.with(holder.binding.tagItemImg.getContext())
                    .load(icon)
                    .error(R.drawable.button_selector_drawable)
                    .into(holder.binding.tagItemImg);
        }

        private int setIcon(boolean isSelected, int position) {
            switch (position) {
                case 0 : {
                    if (isSelected) {
                        return R.drawable.ic_roadtour_on;

                    } else {
                        return R.drawable.ic_roadtour_off;
                    }

                }

                case 1 : {
                    if (isSelected) {
                        return R.drawable.ic_hardtour_on;
                    } else {
                        return R.drawable.ic_hardtour_off;
                    }
                }

                case 2 : {
                    if (isSelected) {
                        return R.drawable.ic_trackingtour_on;
                    } else {
                        return R.drawable.ic_trackingtour_off;
                    }
                }

                case 3 : {
                    if (isSelected) {
                        return R.drawable.ic_festivaltour_on;
                    } else {
                        return R.drawable.ic_festivaltour_off;
                    }

                }

                case 4 : {
                    if (isSelected) {
                        return R.drawable.ic_webtoontour_on;
                    } else {
                        return R.drawable.ic_webtoontour_off;
                    }
                }

                case 5 : {
                    if (isSelected) {
                        return R.drawable.ic_historytour_on;
                    } else {
                        return R.drawable.ic_hardtour_off;
                    }
                }

                default: {
                    return R.drawable.ic_hardtour_off;
                }

            /*case 6 : {
                if (isSelected) {

                } else {

                }
                break;
            }*/
            }
        }

        @Override
        public int getItemCount() {
            return arrayList.size();
        }

        public class ViewHolder extends RecyclerView.ViewHolder {
            DirectionGuidTagItemBinding binding;

            public ViewHolder(@NonNull DirectionGuidTagItemBinding binding) {
                super(binding.getRoot());
                this.binding = binding;
                binding.tagItemTxv.setOnClickListener(v -> {
                    prevSelected = selectedItem;
                    selectedItem = getAbsoluteAdapterPosition();
                    if (getAbsoluteAdapterPosition() != RecyclerView.NO_POSITION) {
                        if (itemOnClick != null) {
                            itemOnClick.onCategoryClick(arrayList.get(getAbsoluteAdapterPosition()));
                            notifyItemChanged(selectedItem);
                            notifyItemChanged(prevSelected);
                        }
                    }
                });

            }
        }
    }
}

File: app/src/main/java/com/sdin/tourstamprally/ui/fragment/report/VisitHistoryFragment.kt
package com.sdin.tourstamprally.ui.fragment.report

import android.annotation.SuppressLint
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.databinding.DataBindingUtil
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.ItemTouchHelper
import androidx.recyclerview.widget.LinearLayoutManager
import com.sdin.tourstamprally.R
import com.sdin.tourstamprally.Utils
import com.sdin.tourstamprally.adapter.report.CategoryAdapter
import com.sdin.tourstamprally.adapter.report.VisitAdapter
import com.sdin.tourstamprally.adapter.swipe.SwipeHelperCallback
import com.sdin.tourstamprally.databinding.FragmentVisithistoryBinding
import com.sdin.tourstamprally.model.CouponModel
import com.sdin.tourstamprally.model.HistorySpotModel
import com.sdin.tourstamprally.model.ReviewWriter
import com.sdin.tourstamprally.model.around.AllAroundEntity
import com.sdin.tourstamprally.ui.dialog.DefaultBSRDialog
import com.sdin.tourstamprally.ui.dialog.DialogEventJoinSuccess
import com.sdin.tourstamprally.ui.dialog.event.TicketDialog
import com.sdin.tourstamprally.ui.fragment.BaseFragment
import com.sdin.tourstamprally.utill.listener.VisitItemClickListener
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.observers.DisposableSingleObserver
import io.reactivex.rxjava3.schedulers.Schedulers
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class VisitHistoryFragment : BaseFragment(), VisitItemClickListener {
    private var binding: FragmentVisithistoryBinding? = null
    private var historySpotList: MutableList<HistorySpotModel>? = null
    private lateinit var visitAdapter: VisitAdapter
    private var currentHistorySpotModel: HistorySpotModel? = null
    private var currentPosition = -1
    private var location = 0
    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding =
            DataBindingUtil.inflate(inflater, R.layout.fragment_visithistory, container, false)
        return binding?.root
    }

    override fun onViewCreated(view: View, savedInstancdState: Bundle?) {
        super.onViewCreated(view, savedInstancdState)

        binding?.apply {
            progressBar.visibility = View.VISIBLE
            setCategory()
            if(location == 0){
                location = 1
                setVisitInit(0)
            }
//            setVisitInit()


            btnGetstamp.setOnClickListener {
                val action = VisitHistoryFragmentDirections.actionPageReportToPageStamp()
                findNavController().navigate(action)
            }


            searchEdt.addTextChangedListener(object : TextWatcher {
                override fun beforeTextChanged(
                    s: CharSequence?,
                    start: Int,
                    count: Int,
                    after: Int
                ) = Unit

                override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) =
                    Unit

                override fun afterTextChanged(s: Editable?) {
                    search(binding?.searchEdt?.text.toString())
                }

            })
        }
    }

    private fun setVisitInit(locationx : Int) {
        apiService.getHistorySpotVisit(Utils.User_Idx,locationx)
            .enqueue(object : Callback<List<HistorySpotModel>> {
                @SuppressLint("ClickableViewAccessibility")
                override fun onResponse(
                    call: Call<List<HistorySpotModel>>,
                    response: Response<List<HistorySpotModel>>
                ) {
                    binding?.progressBar?.visibility = View.GONE
                    if (response.isSuccessful) {

                        val result = response.body()
                        if (result != null) {
                            historySpotList = result.toMutableList()

                            val callback = SwipeHelperCallback()
                            callback.setClamp(-200F)

                            val itemTouchHelper = ItemTouchHelper(callback)
                            itemTouchHelper.attachToRecyclerView(binding?.visitRecyclerView)

                            val list = mutableListOf<HistorySpotModel>()
                            list.addAll(historySpotList!!)

                            visitAdapter = VisitAdapter(reviewWriterCallback = {
                                onWriteReviewClick(it)
                            }, visitItemClickListener = this@VisitHistoryFragment).apply {
                                submitList(list)
                            }

                            binding?.visitRecyclerView?.apply {
                                layoutManager = LinearLayoutManager(requireContext())
                                adapter = visitAdapter
//                                addItemDecoration(ItemDecoration())
                                setOnTouchListener { _, _ ->
                                    callback.removePreviousClamp(this)
                                    false
                                }
                            }

                        }
                    }
                }

                override fun onFailure(call: Call<List<HistorySpotModel>>, t: Throwable) {
                    t.printStackTrace()
                }

            })
    }


    private fun onWriteReviewClick(model: ReviewWriter) {
        val action = VisitHistoryFragmentDirections.actionFragmentVisithistoryToFragmentWriteReview(
            writerModel = model,
            title = "리뷰작성"
        )
        findNavController().navigate(action)
    }


    private fun setCategory() {
        apiService.allLocation.enqueue(object : Callback<List<AllAroundEntity>> {
            override fun onResponse(
                call: Call<List<AllAroundEntity>>,
                response: Response<List<AllAroundEntity>>
            ) {
                val res = response.body()
                res?.let { result ->

                    binding?.categoryRecyclerView?.apply {
                        layoutManager = LinearLayoutManager(requireContext())
                            .apply {
                                orientation = LinearLayoutManager.HORIZONTAL
                            }

                        adapter = CategoryAdapter(callback = {

                            location = it.location_idx

                            setVisitInit(location)

                        }).apply {
                            //setViewType(0)
                            val categoryList = mutableListOf<AllAroundEntity>()
                            val entity = AllAroundEntity(0, "전체보기")
                            categoryList.add(entity)
                            categoryList.addAll(result)
                            //setLocationCount(categoryList.count())
                            submitList(categoryList)
                        }
                    }

                }
            }

            override fun onFailure(call: Call<List<AllAroundEntity>>, t: Throwable) {
                t.printStackTrace()
            }


        })

        val arr: Array<String> = resources.getStringArray(R.array.area_direction)

        binding?.categoryRecyclerView?.apply {
            adapter = CategoryAdapter { location ->
                //var paramData = location
                if (location.location_name == "전체") {
                    //paramData.hashTa = "";
                }
                //search(paramData)
            }.apply {

            }
        }

    }

    private fun getlocationspot(locationidx : Int){

    }

    private fun search(searchData: String) {
        val mutableList = mutableListOf<HistorySpotModel>()
        historySpotList?.let {
            if (searchData.isEmpty()) {
                mutableList.addAll(it)
            } else {

                for (model in it) {
                    if (model.location_name.lowercase().contains(searchData)) {
                        mutableList.add(model)
                    }

                    if (model.touristspot_name.lowercase().contains(searchData)) {
                        mutableList.add(model)
                    }

                }
            }

            visitAdapter.submitList(mutableList)
        }

    }

    private fun showToast(msg: String) {
        Toast.makeText(
            requireContext(), msg, Toast.LENGTH_SHORT
        ).show()
    }

    private fun insertInterest(touristSpotIdx: Int) {
        apiService.insert_intest(Utils.User_Idx, touristSpotIdx).enqueue(object : Callback<Int> {
            override fun onResponse(call: Call<Int>, response: Response<Int>) {
                if (response.isSuccessful) {
                    showToast("찜하기에 성공하였습니다.")
                } else {
                    showToast("찜하기에 실패하였습니다.")
                }
            }

            override fun onFailure(call: Call<Int>, t: Throwable) {
                t.printStackTrace()
                showToast("찜하기에 실패하였습니다.")
            }

        })
    }

    override fun deapClick(position: Int, model: HistorySpotModel) {
        apiService.select_interest_status(Utils.User_Idx, model.touristspot_idx)
            .enqueue(object : Callback<Int> {
                override fun onResponse(call: Call<Int>, response: Response<Int>) {
                    if (response.isSuccessful) {
                        val result = response.body()
                        if (result != null && result != 0) {
                            insertInterest(model.touristspot_idx)
                        }
                    } else {
                        showToast("찜하기에 실패하였습니다.")
                    }
                }

                override fun onFailure(call: Call<Int>, t: Throwable) {
                    showToast("찜하기에 실패하였습니다.")
                    t.printStackTrace()
                }

            })
    }

    override fun deleteReview(reviewIdx: Int, position: Int) {
        apiService.deleteReview(reviewIdx).subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeWith(object : DisposableSingleObserver<Int>() {
                override fun onSuccess(t: Int) {
                    if (t > 0) {
                        visitAdapter.setChange(position = position)
                        showToast("리뷰 삭제에 성공하였습니다.")
                    }

                }

                override fun onError(e: Throwable) {
                    e.printStackTrace()
                }

            })
    }

    override fun clearClick(model: HistorySpotModel, position: Int) {
        currentPosition = position
        currentHistorySpotModel = model

        model.coupon_idx?.let {
            val couponModel = CouponModel(
                it,
                Utils.User_Idx,
                model.coupon_number!!,
                model.coupon_status!!,
                model.touristspot_idx,
                model.coupon_createtime!!,
                model.coupon_createtime!!
            )

            TicketDialog(requireContext(), couponModel, model.couponCount) { pair ->
                if (pair.first == 0) {
                    //다시도전
                    showResetAlertDialog()
                } else if (pair.first == 1) {
                    //응모하기
                    pair.second?.let { paramModel ->
                        joinEvent(paramModel)
                    }
                }
            }.show()
        }


        //PopUp_Image(requireContext(), model).show()
        /*apiService.selectCoupon(tourSpotIdx, Utils.User_Idx).subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread())
            .subscribeWith(object : DisposableSingleObserver<CouponModel>(){
                override fun onSuccess(model : CouponModel) {
                    try {
                        if (model != null) {
                            PopUp_Image(requireContext(), model).show()
                        } else {
                            showToast("해당 관광지의 쿠폰이 발급되지 않았습니다.")
                        }
                    } catch (e : Exception) {
                        e.printStackTrace()
                    }
                }

                override fun onError(e: Throwable) {
                    e.printStackTrace()
                    showToast("해당 관광지의 쿠폰이 발급되지 않았습니다.")
                }

            })*/
    }

    private fun showResetAlertDialog() {
        DefaultBSRDialog(
            context = requireContext(),
            title = "스탬프 기록이 초기화 됩니다.",
            content = "응모하기를 하지 않을 시\n현재 쿠폰으로 응모가 불가합니다.\n다시 도전하시겠습니까?",
            isSpecial = false,
            isSwitchBtn = false,
            leftBtnStr = "도전",
            rightBtnStr = "응모하기"
        ) { callback ->

            if (callback) {
                //초기화

                if (currentHistorySpotModel != null) {
                    apiService.resetTourSpot(
                        currentHistorySpotModel!!.touristspot_idx,
                        Utils.User_Idx
                    )
                        .subscribeOn(Schedulers.newThread())
                        .observeOn(AndroidSchedulers.mainThread())
                        .subscribeWith(object : DisposableSingleObserver<Int>() {
                            override fun onSuccess(result: Int) {
                                if (result > 0) {
                                    // 삭제 성공

                                    val list = visitAdapter.currentList.toMutableList()
                                    list.removeAt(currentPosition)
                                    visitAdapter.submitList(list)

                                    currentHistorySpotModel = null
                                }
                            }

                            override fun onError(e: Throwable) {
                                e.printStackTrace()
                            }
                        })
                }

            }
            else{
                //joinEvent

                currentHistorySpotModel?.coupon_idx?.let {
                    val couponModel = CouponModel(
                        it,
                        Utils.User_Idx,
                        currentHistorySpotModel?.coupon_number!!,
                        currentHistorySpotModel?.coupon_status!!,
                        currentHistorySpotModel?.touristspot_idx!!,
                        currentHistorySpotModel?.coupon_createtime!!,
                        currentHistorySpotModel?.coupon_createtime!!
                    )

                    joinEvent(couponModel)

                }
            }

        }.show()
    }

    private fun showSuccessEventJoin() {
        DialogEventJoinSuccess(requireContext()) {
            requireActivity().startActivity(
                Intent(
                    Intent.ACTION_VIEW, Uri.parse("http://bsrp.co.kr/list.php?bbs_id=main")
                )
            )

        }.show()
    }

    private fun joinEvent(model: CouponModel) {
        model.coupon_touristspot_idx
        if (model.coupon_status > 0) {
            //참여 완료 팝업
            showSuccessEventJoin()

        } else {
            apiService.insertEvent(Utils.User_Idx, model.coupon_idx, model.coupon_number)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribeWith(object : DisposableSingleObserver<Int>() {
                    override fun onSuccess(result: Int) {
                        when {
                            result >= 0 -> {
                                showSuccessEventJoin()
                            }
                            result > 0 -> {
                                //참여 완료 팝업
                                showSuccessEventJoin()
                            }
                            else -> {
                                //이벤트 창여 실패
                            }
                        }
                    }

                    override fun onError(e: Throwable) {
                        e.printStackTrace()
                    }
                })
        }

    }


}